Fixed midiCC MSB message handling including UI updates

Vlietland opened the pull request:
Dear Jose, herewith my source code changes to enable midicc MSB handling. Since I am new in this repo, please check my changes thoroughly, to exclude possible introduced bugs. 

Many thanks and regards in advance for checking and hopefully accepting my changes. 

Jan

Commit:
Fixed midiCC MSB message handling including UI updates

jofemodo commented on zyngui/zynthian_gui_layer.py:
```diff
@@ -507,8 +507,9 @@ def set_midi_chan_preset(self, midich, preset_index):
-						if not self.zyngui.modal_screen and self.zyngui.active_screen in ('control'):
-							self.select_action(self.root_layers.index(layer))
+						#following conditional statement removed to omit issue with not updating the screen
+						#if not self.zyngui.modal_screen and self.zyngui.active_screen in ('control'):
```
We don't want preset loading to change the active screen. What you see as a bug, it's a feature ;-)
Or perhaps i'm not understanding the problem you have.

Vlietland replied:
I proposed this change to omit the issue that the screen is not updated after processing the midi bank change message and a program control change message (instrument selection). For instance when the bank is updated the bank screen data shown is not changed. Same goes for the preset screen and likely also for the layer screen. This is very confusing for the user.

Do you have a low impact alternative solution to update the presently selected screen.

jofemodo commented on zyngine/zynthian_layer.py:
```diff
@@ -1,13 +1,13 @@
 #******************************************************************************
 # ZYNTHIAN PROJECT: Zynthian Layer (zynthian_layer)
-# 
+#
```
Please, remove all "code formating" changes from the commit.
If you want to improve code format, then make a separated commit with the "code formatting" changes.

jofemodo commented:
Ahhh! Please, always PR to "testing" branch, never to "stable" branch.

Thanks

jofemodo commented on zyngine/zynthian_layer.py:
```diff
@@ -395,8 +395,16 @@ def midi_unlearn(self):
+			#logging.debug("Setting Preset: Receiving MIDI CH{}#CC{}={}".format(chan, ccnum, ccval))
+			pass
+		elif self.engine:
+			#logging.debug("Receiving MIDI CH{}#CC{}={}".format(chan, ccnum, ccval))
```
I propose this behaviour:
+ When CC0 (MSB) is received, we save it to a flag var "bank_msb". 
+ When CC32 (LSB) is received, we change the bank, using the "bank_msb" like this:

    + bank_msb==0 => system banks LSB (0-127)
    + bank_msb==1 => user bank LSB (0-127)
    + bank_msb==2 => favourites bank, if it exist. In this case, lsb has not meaning.

Of course, other schemes are possible, but i find this very simple and  easy to remember/understand.

jofemodo commented:
Closing this. I just implemented the feature in a separated branch using your initial idea & code:

https://github.com/zynthian/zynthian-ui/tree/midi_bank_change

It implements MSB like this:

+ 0 => system banks
+ 1 => user banks
+ 2 => external banks (USB drive)

LSB is the bank index. 

I improved the "refresh GUI" code, so now it should work fine. For instance, if active screen is "layers", it will get refreshed on bank/preset changes. If active screen is "bank" or "preset" list, it will jump to control. Etc.

Thanks a lot!

Vlietland commented:
You are conveniently fast Jose. I had already implemented most of the MSB/LSB code myself and wanna to upload the solution to the test branch tonight. I will take a look at it and see if we implemented the same type of solution (for learning purposes).
